docs/bsp.md

```diff
@@ -902,14 +1072,63 @@ trait TestReport {
 }
 ```
 
-The `TestReport` notification will be sent as the test execution of build targets completes.
+This request may trigger a compilation or other tasks on the selected build targets.
 
-This request may trigger a compilation on the selected build targets. The server is free to send any
-number of `build/compileReport`, `build/publishDiagnostics` and `build/logMessage` notifications
-during compilation before completing the response.
+The server may send any number of `build/task*`, `build/publishDiagnostics` and `build/logMessage` notifications
+to communicate about tasks triggered by the request to the client.
+
+#### Test Unit Notifications
 
-The client will get a `originId` field in the `TestReport` or `TestResult` if the `originId` field
-in the `TestParams` is defined.
+The server may inform about individual test units in task notifications that reference the originating task in their `taskId`.
+For example, the server can send a `taskStart`/`taskFinish` for each test suite in a target, and likewise for each test in the suite. 
+Individual test start notifications should specify `test-started` in the `dataKind` field and include the `TestStart` object and
+test finish notifications should specify `test-finished` in the `dataKind` field and include the `TestFinish` object in the `data` field.
+
+```scala
+trait TestStart {
+  /** Name or description of the test. */
+  def description: String
+  
+  /** Source location of the test, as LSP location. */
+  def location: Option[Location]
+}
+
+trait TestFinish {
```

Comment:
I don't think anything about this suggests that these notifications are compulsory. The build tool can implement them at whatever granularity is possible or practical.

For example, if it only has information about all the tests in a suite at a time, it could report a `TestFinish` for each test once the suite is done. If it can report only on the suite as a whole, the same `TestFinish` can be used or just use plain messages with `StatusCode`

Reply:
> I don't think anything about this suggests that these notifications are compulsory.
  
The documentation of the fields mislead me to think it's compulsory to report on a test-by-test basis, claiming that they are metadata about "tests" only.
  
> For example, if it only has information about all the tests in a suite at a time, it could report a TestFinish for each test once the suite is done. 
  
If this is the case, I propose:
  
1. We explicitly mention that the build tool can use the granularity it wants.
2. To rename the documentation to not be tied to a concrete test.

Reply:
Agreed